sdk/core/azure-core/src/main/java/com/azure/core/implementation/ResponseConstructorsCache.java

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.implementation;
+
+import com.azure.core.http.HttpHeaders;
+import com.azure.core.http.HttpRequest;
+import com.azure.core.http.HttpResponse;
+import com.azure.core.http.rest.Response;
+import com.azure.core.implementation.serializer.HttpResponseDecoder;
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.Exceptions;
+import reactor.core.publisher.Mono;
+
+import java.lang.invoke.LambdaMetafactory;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.reflect.Constructor;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * A concurrent cache of {@link Response} constructors.
+ */
+final class ResponseConstructorsCache {
+    private final ClientLogger logger = new ClientLogger(ResponseConstructorsCache.class);
+    private final Map<Class<?>, Optional<ResponseConstructor>> cache = new ConcurrentHashMap<>();
+
+    /**
+     * Identify the most specific constructor for the given response class.
+     *
+     * The most specific constructor is looked up following order:
+     * 1. (httpRequest, statusCode, headers, body, decodedHeaders)
+     * 2. (httpRequest, statusCode, headers, body)
+     * 3. (httpRequest, statusCode, headers)
+     *
+     * @param responseClass the response class
+     * @return optional with located constructor, empty optional if there is no match
+     */
+    Optional<ResponseConstructor> get(Class<? extends Response<?>> responseClass) {
+        this.cache.computeIfAbsent(responseClass, responseCls -> Arrays.stream(responseCls.getDeclaredConstructors())
+            .filter(constructor -> {
+                int paramCount = constructor.getParameterCount();
+                return paramCount >= 3 && paramCount <= 5;
+            })
+            .sorted(Comparator.comparingInt(Constructor::getParameterCount))
+            .findFirst()
+            .flatMap(constructor -> {
+                try {
+                    MethodHandles.Lookup lookup = MethodHandles.lookup();
+                    MethodHandle ctrMethodHandle = lookup.unreflectConstructor(constructor);
+                    switch (constructor.getParameterCount()) {
+                        case 3:
+                            Object f3 = LambdaMetafactory.metafactory(lookup,
```

Comment:
Can the type of `f3` be more specific than `Object`?

Reply:
LambdaMetafactory 's `.invoke()` returns `Object` (this method is not generic) and this Object is of `ResponseFunc3`.

We could cast it to the specific type `ResponseFunc3` but our cache value is of type `Object`. We do casting when we use it to invoke ctr.

I was initially worried about usage of `Object` but I'm seeing this is happening quite often when working with low level LambdaMetaFactory.